src/sif/autocost.cc

```diff
@@ -408,32 +425,28 @@ Cost AutoCost::EdgeCost(const baldr::DirectedEdge* edge,
   // either the computed edge speed or optional top_speed
   auto edge_speed = tile->GetSpeed(edge, flow_mask_, seconds);
   auto final_speed = std::min(edge_speed, top_speed_);
-
-  float sec = (edge->length() * speedfactor_[final_speed]);
+  float sec = edge->length() * speedfactor_[final_speed];
 
   if (shortest_) {
     return Cost(edge->length(), sec);
   }
 
-  float factor =
-      (edge->use() == Use::kFerry)
-          ? ferry_factor_
-          : (edge->use() == Use::kRailFerry) ? rail_ferry_factor_ : density_factor_[edge->density()];
+  // base factor is either ferry, rail ferry or density based
+  float factor = (edge->use() == Use::kFerry) * ferry_factor_ +
+                 (edge->use() == Use::kRailFerry) * rail_ferry_factor_ +
+                 (edge->use() != Use::kFerry && edge->use() != Use::kRailFerry) *
+                     density_factor_[edge->density()];
 
-  // TODO: factor hasn't been extensively tested, might alter this in future
+  // TODO: speed_penality hasn't been extensively tested, might alter this in future
   float speed_penalty = (edge_speed > top_speed_) ? (edge_speed - top_speed_) * 0.05f : 0.0f;
-  factor += highway_factor_ * kHighwayFactor[static_cast<uint32_t>(edge->classification())] +
-            surface_factor_ * kSurfaceFactor[static_cast<uint32_t>(edge->surface())] + speed_penalty;
-
-  if (edge->toll()) {
-    factor += toll_factor_;
-  }
-
-  if (edge->use() == Use::kAlley) {
-    factor *= alley_factor_;
-  }
-
-  return Cost(sec * factor, sec);
+  factor += (highway_factor_ * kHighwayFactor[static_cast<uint32_t>(edge->classification())] +
+             surface_factor_ * kSurfaceFactor[static_cast<uint32_t>(edge->surface())] +
+             speed_penalty + edge->toll() * toll_factor_) *
+            (edge->use() == Use::kAlley ? alley_factor_ : 1.f);
```

Comment:
i removed all the branches except 1 in favor of a multiplcation trick

Reply:
"trick" == using a boolean as a double?  :).  I wonder if that generates a compiler warning...

Reply:
If this is a commonly used convention that simplifies the logic throughout, I can learn to live with promoting bools to doubles (or ints).  Just an observation.

Reply:
bool gets upcasted to float yes. since float has more precision than bool there is no warning on standard warning levels though `-Wall` or `-Wpedantic` may warn about this i havent checked. we do have `-Werror` enabled for `sif` in the build so we are fine there for the moment imho

Reply:
locally ive add both `-Wall` and `-Wpedantic` and indeed it doesnt complain (though pedantic did find an extra semicolon in some other code :wink:

Reply:
@ktatterso it *is* commonly used in this codebase, but I tend to agree with you - it makes for terser code, but I don't think that always means simpler logic or improved readability.

I would like to see a benchmark to prove out that removal of branches here is actually significant.  I know the *theory* is to reduce branching -> improved pipelining -> faster code, but I have my doubts that these kinds of branches actually end having a significant impact in many places.

Reply:
The compiler _should_ be able to optimize the simple branches away

Reply:
yeah i recently made some performance improvements to isochrones that resulted in something like a 10% gain by removing a cascade of if elses. we got a smaller gain by reducing branching and recursion in bidirectional_astar as well. i just wanted to do an experiment here to see if we could get anything out of it.

the compiler will likely be able to predict a lot of these that occur with relatively low frequency i agree. in c++20 we can even influence that with the `unlikely` keyword

Reply:
@PureW I'm not sure the compiler can optimize away any condition unless it is programmatically impossible, e.g., "if ( false )".  Maybe you can clarify?

Compiler optimized branch "prediction" can be baked into our binaries if we can employ some sort of "profile guided optimization".  I can speak to this a bit further if there is interest.

In the end tho, the switch statement is readable (imo), doesn't rely on compiler optimization or branch prediction of any sort, and provably faster than cascading conditionals (tho as @danpat points out, not always measurably faster).